Documentation/General/finalReport/reflection.tex

```diff
@@ -1 +1,107 @@
-\chapter{Reflection on the product and process from a software engineering perspective}
\ No newline at end of file
+\chapter{Reflection on the product and process from a software engineering perspective}
+In this chapter, we discuss the software engineering perspectives of MoodCat. First, we discuss our developing process. Second, we discuss the general structure and architecture of the product. Then we dive a bit deeper into the design paterns we use and considered. We also touch on code readability and maintainability, and the choices we made there. Finally, we list our most important libraries and languages used.
+
+\section{Product Architecture}
+The basic model of MoodCat's architecture is the same as just about every other website.
+We're using a client-server architecture.
+We actually do this in the form of an AngularJS web application and a Java API based on JAX-RS\footnote{The implementation we use is RestEASY, which can be found on \url{http://resteasy.jboss.org/}}. 
+The API in turn talks to a persistence server in the form of a PostgreSQL database.
+
+\section{Design Patterns}
+A lot of the design patterns in our application come from the libraries and frameworks we use. However, we did design our backend system very methodically, using the model-view-controller (MVC) pattern. The API classes define the view of the application, in the form of JSON responses, which is what our frontend uses to get it's data. The "backend" package is what defines the controller of the system, and the DAO's (Data Access Object) define our models. They talk to the postgreSQL database for retrieval and persistence.
+
+In the frontend we use a slightly different structure, most commonly referred to MVVM (Model, View, ViewModel). The models here are the API calls, controllers and services are the viewmodels, and the HTML templates are the views.
+
+A few of the design patterns we used:
+\begin{enumerate}
+\item facade
+\item proxy
+\item singleton
+\item factories
+\item Dependency injection
+\item MVC
+\item MVVM
+\item Resource pool
+\end{enumerate}
+
+\section{Code readability / maintainability}
+One of our goals in this project was creating a readable and maintainable codebase.
```

Comment:
Imo Lombok is mostly a development efficiency improver, and the code readability benefit is quite minor. While our system is really developed with maintainability and readability in mind, I think we should focus on a few other properties in this paragraph:
* We use small method bodies with descriptive names
* We use lamba's and streams, which are way more readable than loops, nested ifs, unnecessary copies and above all, anonymous classes.
* We use QueryDSL to write queries using a builder pattern, which makes queries more readable and adds the benefit of type checking for our queries.
* We use Google Guice which really allows us to rapid prototype classes and use composition at its best and move dependencies around in refactoring.
* We use various contracts and classes for the interactions between the database layer, backend layer and API layer, which thus can be seen as a layered model, which add the benefit that these layers can be individually maintained while the other layers will keep working.